haystack/retriever/elasticsearch.py

```diff
@@ -41,12 +41,58 @@ def __init__(self, document_store: Type[BaseDocumentStore], custom_query: str =
         self.document_store = document_store
         self.custom_query = custom_query
 
-    def retrieve(self, query: str, filters: dict = None, top_k: int = 10) -> [Document]:
-        documents = self.document_store.query(query, filters, top_k, self.custom_query)
+    def retrieve(self, query: str, filters: dict = None, top_k: int = 10, index: str = None) -> [Document]:
+        if index is None:
+            index = self.document_store.index
+
+        documents = self.document_store.query(query, filters, top_k, self.custom_query, index)
         logger.info(f"Got {len(documents)} candidates from retriever")
 
         return documents
 
+    def eval(self, label_index: str = "feedback", doc_index: str = "eval_document", label_origin: str = "gold_label",
+             top_k: int = 10) -> dict:
+        """
+        Performs evaluation on the Retriever.
+        Retriever is evaluated based on whether it finds the correct document given the question string and at which
+        position in the ranking of documents the correct document is.
+
+        Returns a dict containing the following metrics:
+            - "recall": Proportion of questions for which correct document is among retrieved documents
+            - "mean avg precision": Mean of average precision for each question. Rewards retrievers that give relevant
+              documents a higher rank.
+
+        :param label_index: Elasticsearch index where labeled questions are stored
```

Comment:
While Elasticsearch is currently our recommended DocumentStore we should phrase this a bit more generic (e.g. "Index/Table in DocumentStore where labeled questions are stored"). If I got it right supporting eval for another DocumentStore only requires implementations of the methods "add_eval_data()" and "get_all_docs_in_index()".